modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala

```diff
@@ -130,16 +160,37 @@ final case class Resolver[F[_]](repos: List[Registry], cache: Option[ResolverCac
   )(implicit
     F: Monad[F],
     C: Clock[F]
-  ): F[Option[SchemaLookup]] =
+  ): F[Option[ResolverCache.TimestampedItem[SchemaLookup]]] =
     cache match {
-      case Some(c) => c.getSchema(schemaKey)
+      case Some(c) => c.getTimestampedSchema(schemaKey)
       case None    => Monad[F].pure(None)
     }
 }
 
 /** Companion object. Lets us create a Resolver from a Json */
 object Resolver {
 
+  /** The result of doing a lookup with the resolver, carrying information on whether the cache was used */
+  sealed trait ResolverResult[+A] {
+    def value: A
+  }
+
+  /**
+   * The result of a lookup when the resolver is configured to use a cache
+   *
+   *  The timestamped value is helpful when the client code needs to perform an expensive
+   *  calculation derived from the looked-up value. If the timestamp has not changed since a
+   *  previous call, then the value is guaranteed to be the same as before, and the client code
+   *  does not need to re-run the expensive calculation.
+   *
+   *  @param value the looked-up value
+   *  @param timestamp epoch time in seconds of when the value was last cached by the resolver
+   */
+  case class Cached[A](key: SchemaKey, value: A, timestamp: Int) extends ResolverResult[A]
```

Comment:
Do we need type param `A`? We have `SchemaKey` here so we're committed to schema. And schema == `Json` value, right?

Reply:
And if it's only about schema: change name to `SchemaLookupResult`?

Reply:
Aaaand make the trait private?

Reply:
The reason for type param `A`.... it's because `Resolver` has two main methods, `lookupSchema` and `listSchemas`.  The idea was to return `Cached[Json]` for a single lookup, and `Cached[SchemaList]` for the listing.

We should review our snowplow apps (mainly rdb transformer in shredding mode) and see if we might ever need the cache timestamp on `SchemaList`.  If we will never need a `Cached[SchemaList]` then I agree we don't need the type parameter.

>  change name to SchemaLookupResult?

Similarly to the above, I named it `ResolverResult` in case it might be useful for multiple types of lookup.  I guess we could alias it `type SchemaLookupResult = ResolverResult[Json]`.  Or I'm also happy to rename it with your suggestion if you think the lookup-flexibility is not needed.

> make the trait private?

This was deliberately public.  I was thinking for Snowplow loaders the logic would be:

- Use the resolver to lookup a `ResolverResult[Json]`
- Use a LruMap to lookup the schema-ddl column type.  For example, BigQuery loader will have a `LruMap[F, (SchemaKey, Int), Field`, to lookup [a BigQuery Field](https://github.com/snowplow/schema-ddl/blob/0.15.0/modules/core/src/main/scala/com.snowplowanalytics/iglu.schemaddl/bigquery/Field.scala)
- If it's missing from the LruMap then parse the schema and build the `Field`, which is a cpu-expensive operation.

You've already reviewed my [attempt to get this working in RDB Transformer](https://github.com/snowplow/snowplow-rdb-loader/pull/973), so you'll know I don't have it working yet as I described.  Currently in the transformer [my LruMap](https://github.com/snowplow/snowplow-rdb-loader/pull/973/files#diff-ead9dcf097f06399c326cb8072059effcf4aa4100555c876651acc0a8f7e1efcR28) is keyed by `SchemaKey`, not by `(SchemaKey, Int)`.

Our next step could be to spike solutions in RDB Transformer and BigQuery Loader, and then revisit some of these decisions here before publishing.